### DISCUSSION TITLE
Trying to understand socket, @live and refetch logic

### DISCUSSION THREAD

**Discussion by strblr** [n1ru4l/graphql-live-query#768] [Q&A] [answered]
Hi,

I made some tests on my app using your Socket.IO server and client, and `urql`. I wanted to know what happens when the socket of a user A disconnects (let's say he loses wifi for 3 seconds), then an update is made on the server from another user B, then user A reconnects automatically (Socket.IO does that). Does the reconnection trigger a refetch and a UI sync ?

Turns out it depends on whether the `@live` directive is present or not. If it is present, the UI resyncs on reconnection, otherwise it does not.

I have no idea what's going on in the background that makes that happen. My intuition tells me that the server loses completely track of user A if the socket connection is lost, thus deleting every server state about him. Then on reconnection, entire new queries have to be made. Why does the behavior change depending on the presence of the `@live` directive ?

Thanks !

**Comment by n1ru4l** [comment, marked as answer]
An operation that is finished is disposed of by the `SocketIOGraphQLClient`. A "normal" mutation and query are simply executed and after the last result comes in the `SocketIOGraphQLClient` client deletes any reference to it and is no longer aware of it. This is basically the behavior of all GraphQL clients out there today. 

See https://github.com/n1ru4l/graphql-live-query/blob/c77923de7696e9910f28004babe0fcf9666d484e/packages/socket-io-graphql-client/src/createSocketIOGraphQLClient.ts#L29-L40

Live queries never have an `isFinal == true` property, thus they are never disposed of and re-executed after the network connection got interrupted. The same applies to any other operation for which the client never received a result before the network interruption. As soon as the network connection is interrupted the server disposes of all information about a client.

In order to avoid Mutations being executed twice, I recommend sending a client-generated transaction-id that can be used for identifying whether a mutation operation got executed before, but failed to deliver the mutation execution result to the client due to the network interruption.

**Reply by strblr** [reply]
Makes sense, thanks!

**Reply by n1ru4l** [reply]
BTW, I just want to note that you must not use my GraphQL over Socket.io protocol. The [graphql-ws protocol](https://github.com/enisdenjo/graphql-ws) works with live queries as well and is probably more battle-tested.

**Reply by strblr** [reply]
> BTW, I just want to note that you must not use my GraphQL over Socket.io protocol. The [graphql-ws protocol](https://github.com/enisdenjo/graphql-ws) works with live queries as well and is probably more battle-tested.

I kinda like that Socket.IO provides automatic fallbacks and has very easy reconnection mechanisms. How would you implement the `extensions` feature with ws ?
